runtime/j9vm/asgct.cpp

```diff
@@ -0,0 +1,158 @@
+ticks_GC_active             = -2, // GC action
+ticks_unknown_not_Java      = -3, // ¯\_(ツ)_/¯
+ticks_not_walkable_not_Java = -4, // ¯\_(ツ)_/¯
+ticks_unknown_Java          = -5, // ¯\_(ツ)_/¯
+ticks_not_walkable_Java     = -6, // ¯\_(ツ)_/¯
+ticks_unknown_state         = -7, // ¯\_(ツ)_/¯
+ticks_thread_exit           = -8, // dying thread
+ticks_deopt                 = -9, // mid-deopting code
+ticks_safepoint             = -10 // ¯\_(ツ)_/¯
+}; 
+
+typedef struct {
+	jint lineno;
+	jmethodID method_id;
+} ASGCT_CallFrame;
+
+typedef struct {
+	JNIEnv *env_id;
+	jint num_frames;
+	ASGCT_CallFrame *frames;
+} ASGCT_CallTrace;
+
+#if defined(ASGCT_SUPPORTED)
+
+extern J9JavaVM *BFUjavaVM;
+
+static UDATA
+asyncFrameIterator(J9VMThread * currentThread, J9StackWalkState * walkState)
+{
+	ASGCT_CallFrame *frame = (ASGCT_CallFrame*)walkState->userData1;
+	J9Method *method = walkState->method;
+	J9JNIMethodID *methodID = currentThread->javaVM->internalVMFunctions->getJNIMethodID(currentThread, method);
+	Assert_SC_notNull(methodID);
+	frame->method_id = (jmethodID)methodID;
+	J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+	if (J9_ARE_ANY_BITS_SET(romMethod->modifiers, J9AccNative)) {
+		frame->lineno = -3;
+	} else {
+		frame->lineno = (jint)walkState->bytecodePCOffset;
+	}
+	walkState->userData1 = (void*)(frame + 1);
+	return J9_STACKWALK_KEEP_ITERATING;
+}
+
+static UDATA
+emptySignalHandler(J9PortLibrary *portLibrary, U_32 gpType, void *gpInfo, void *handler_arg)
+{
+   return J9PORT_SIG_EXCEPTION_RETURN;
+}
+
+typedef struct {
+	ASGCT_CallTrace *trace;
+	jint depth;
+	void *ucontext;
+	J9VMThread *currentThread;
+	jint num_frames;
+} ASGCT_parms;
+
+static UDATA
+protectedASGCT(J9PortLibrary *portLib, void *arg)
+{
+	ASGCT_parms *parms = (ASGCT_parms*)arg;
+	J9VMThread *currentThread = BFUjavaVM->internalVMFunctions->currentVMThread(BFUjavaVM);
+	if (NULL != currentThread) {
+		parms->currentThread = currentThread;
+		parms->num_frames = ticks_not_walkable_Java;
+		J9JITConfig *jitConfig = BFUjavaVM->jitConfig;
+		if (NULL != jitConfig) {
+			void *ucontext = parms->ucontext;
+			if (NULL != ucontext) {
+				greg_t *regs = ((ucontext_t*)ucontext)->uc_mcontext.gregs;
+				greg_t rip = regs[REG_RIP];
+				J9JITExceptionTable *metaData = jitConfig->jitGetExceptionTableFromPC(currentThread, rip);
+				if (NULL != metaData) {
+					greg_t rsp = regs[REG_RSP];
+					jitPushResolveFrame(currentThread, (UDATA*)rsp, (U_8*)rip);
+				}
+			}
+		}
+		J9StackWalkState walkState;
+		walkState.walkThread = currentThread;
+		walkState.skipCount = 0;
+		walkState.maxFrames = parms->depth;
+		walkState.flags = J9_STACKWALK_INCLUDE_NATIVES | J9_STACKWALK_VISIBLE_ONLY
+			| J9_STACKWALK_RECORD_BYTECODE_PC_OFFSET | J9_STACKWALK_COUNT_SPECIFIED
+			| J9_STACKWALK_ITERATE_FRAMES;
+		walkState.userData1 = (void*)parms->trace->frames;
+		walkState.frameWalkFunction = asyncFrameIterator;
+		UDATA result = BFUjavaVM->walkStackFrames(currentThread, &walkState);
```

Comment:
is it possible for an allocation to occur here? ie in `walkBytecodeFrameSlots`?

not sure we can do that in signal handler code

Reply:
We aren't iterating slots, so that code will not be executed.

Reply:
There is also the artifact search cache, but let's deal with that in a subsequent PR.